core_api/src/routes/chat.py

```diff
@@ -162,3 +168,125 @@ def rag_chat(chat_request: ChatRequest, _user_uuid: Annotated[UUID, Depends(get_
+async def websocket_endpoint(websocket: WebSocket):
+    await websocket.accept()
+
+    prompt_search_query = ChatPromptTemplate.from_messages(
+        [
+            MessagesPlaceholder(variable_name="chat_history"),
+            ("user", "{input}"),
+            (
+                "user",
+                "Given the above conversation, generate a search query to look up to get information relevant to the conversation",
+            ),
+        ]
+    )
+
+    retriever_chain = create_history_aware_retriever(llm, vector_store.as_retriever(), prompt_search_query)
+
+    prompt_get_answer = ChatPromptTemplate.from_messages(
+        [
+            ("system", "Answer the user's questions based on the below context:\\n\\n{context}"),
+            MessagesPlaceholder(variable_name="chat_history"),
+            ("user", "{input}"),
+        ]
+    )
+
+    document_chain = create_stuff_documents_chain(llm, prompt_get_answer)
+    retrieval_chain = create_retrieval_chain(retriever_chain, document_chain)
+
+    while True:
+        chat_request = ChatRequest.parse_raw(await websocket.receive_text())
+        chat_history = [
+            HumanMessage(content=x.text) if x.role == "user" else AIMessage(content=x.text)
+            for x in chat_request.message_history[:-1]
+        ]
+        chat = {"chat_history": chat_history, "input": chat_request.message_history[-1].text}
+
+        async for event in retrieval_chain.astream_events(chat, version="v1"):
+            kind = event["event"]
+            if kind == "on_chat_model_stream":
+                await websocket.send_text(event["data"]["chunk"].content)
+            if kind == "on_chat_model_end":
+                await websocket.send_text("</chat>")
+            elif kind == "on_retriever_end":
+                source_documents = [
+                    SourceDocument(
+                        page_content=document.page_content,
+                        file_uuid=document.metadata.get("parent_doc_uuid"),
+                        page_numbers=document.metadata.get("page_numbers"),
+                    )
+                    for document in event["data"]["output"]["documents"]
+                ]
+                source_documents_bytes = SourceDocuments(source_documents=source_documents).model_dump_json().encode()
+                await websocket.send_bytes(source_documents_bytes)
+
+
+html = """
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>Chat</title>
+        <script>
+            var ws = new WebSocket("ws://localhost:5002/chat/rag-stream");
+
+            // Change binary type from "blob" to "arraybuffer"
+            ws.binaryType = "arraybuffer";
+
+            var decoder = new TextDecoder("utf-8");
+
+            // Listen for messages
+            ws.addEventListener("message", (event) => {
+              if (event.data instanceof ArrayBuffer) {
+                document.getElementById('documents').innerHTML = JSON.stringify(JSON.parse(decoder.decode(event.data)),null,2);
```

Comment:
Do we want to be keeping a record of the sources throughout the chat rather than just the most recent?  If so, would this be embedded with the relevant chat answer or all in one block at the end?